### PR TITLE
(INFC-17763) Support Windows

### PR DESCRIPTION
**Pull request by danielparks**
This refactors the code and makes extensive changes to support both [native][] and Cygwin OpenSSH on Windows.

All that is required to release 1.0 is documentation, so this bumps the version to 0.9.0.

This is built on top of https://github.com/ploperations/ploperations-ssh/pull/4 and https://github.com/ploperations/ploperations-ssh/pull/3.

[native]: https://github.com/PowerShell/openssh-portable

### CODE REVIEW

File: manifests/params.pp [hunk 1]
```diff
@@ -3,68 +3,173 @@
...
+        $package_provider  = 'cygwin'
+        $config_dir        = cygwin::windows_path('/etc')
+        $manage_config_dir = false
+        $sftp_subsystem    = '/usr/bin/sftp-server'
+        $authorized_keys   = '/etc/user_authorized_keys/%u'
+        $strict_modes      = false # I couldn't figure these out
+        $dos_line_endings  = false
+      } else {
+        # Use native openssh
+        $_programdata = $facts.get('windows_env.PROGRAMDATA')
```
Comment (genebean) [inline manifests/params.pp]: `facts.get()` doesn't work pre-puppet 6

File: manifests/server/windows/chocolatey.pp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Configuration for native OpenSSH
+class ssh::server::windows::chocolatey (
+  Enum[present, absent] $default_shell_ensure         = present,
+  Stdlib::Absolutepath  $default_shell                = "${facts.get('windows_env.WINDIR')}\\System32\\WindowsPowerShell\\v1.0\\powershell.exe",
```
Comment (genebean) [inline manifests/server/windows/chocolatey.pp]: `facts.get()` doesn't work pre-puppet 6

File: manifests/server/windows/cygwin.pp [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# SSH daemon configuration specific to Cygwin
+class ssh::server::windows::cygwin (
+  Sensitive[String[1]] $cyg_server_password,
+) {
+  user { 'cyg_server':
+    comment  => 'Privilege separation for Cygwin sshd',
+    password => $cyg_server_password,
+  }
+
+  $windir = $facts.get('windows_env.WINDIR')
```
Comment (genebean) [inline manifests/server/windows/cygwin.pp]: `facts.get()` doesn't work pre-puppet 6

**Review by genebean — CHANGES_REQUESTED** [review]

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -26,11 +26,23 @@
 end
 
 RSpec.configure do |c|
+  c.mock_with :rspec do |mocks|
+    # Necessary to use allow_any_instance_of
+    mocks.syntax = [:expect, :should]
+  end
+
```
Comment (danielparks) [inline spec/spec_helper.rb:33]: These will be overwritten by PDK — see https://github.com/puppetlabs/pdk-templates#specspec_helperrb.

**Comment by genebean** [comment]
@danielparks are you ready for another review yet or is this still a WIP?

**Comment by danielparks** [comment]
I'll reopen this when it's ready.

**Pull request closed without merging**